src/data/map.rs

```diff
@@ -0,0 +1,1561 @@
+
+/// Public + Sentried
+impl MapBase<PublicAccessList, Sentried> {
+    /// Returns new instance of private MapBase flavour with concurrency control.
+    pub fn new(name: XorName, tag: u64) -> Self {
+        Self {
+            address: Address::PublicSentried { name, tag },
+            data: BTreeMap::new(),
+            access_list: Vec::new(),
+            owners: Vec::new(),
+            version: Some(0),
+            _flavour: Sentried,
+        }
+    }
+}
+
+impl Debug for MapBase<PublicAccessList, Sentried> {
+    fn fmt(&self, formatter: &mut Formatter) -> fmt::Result {
+        write!(formatter, "PublicSentriedMap {:?}", self.name())
+    }
+}
+
+/// Public + NonSentried
+impl MapBase<PublicAccessList, NonSentried> {
+    /// Returns new instance of public MapBase flavour.
+    pub fn new(name: XorName, tag: u64) -> Self {
+        Self {
+            address: Address::Public { name, tag },
+            data: BTreeMap::new(),
+            access_list: Vec::new(),
+            owners: Vec::new(),
+            version: None,
+            _flavour: NonSentried,
+        }
+    }
+}
+
+impl Debug for MapBase<PublicAccessList, NonSentried> {
+    fn fmt(&self, formatter: &mut Formatter) -> fmt::Result {
+        write!(formatter, "PublicMap {:?}", self.name())
+    }
+}
+
+/// Private + Sentried
+impl MapBase<PrivateAccessList, Sentried> {
+    /// Returns new instance of private MapBase flavour with concurrency control.
+    pub fn new(name: XorName, tag: u64) -> Self {
+        Self {
+            address: Address::PrivateSentried { name, tag },
+            data: BTreeMap::new(),
+            access_list: Vec::new(),
+            owners: Vec::new(),
+            version: Some(0),
+            _flavour: Sentried,
+        }
+    }
+
+    /// Commit transaction.
+    ///
+    /// If the specified `expected_version` does not equal the entries count in data, an
+    /// error will be returned.
+    pub fn hard_commit(&mut self, tx: &SentriedTransaction) -> Result<()> {
+        // Deconstruct tx into inserts, updates, and deletes
+        let operations: SentriedOperations = tx.iter().fold(
+            Default::default(),
+            |(mut insert, mut update, mut delete), cmd| {
+                match cmd {
+                    SentriedCmd::Insert(sentried_kvpair) => {
+                        let _ = insert.insert(sentried_kvpair.clone());
+                    }
+                    SentriedCmd::Update(sentried_kvpair) => {
+                        let _ = update.insert(sentried_kvpair.clone());
+                    }
+                    SentriedCmd::Delete(sentried_key) => {
+                        let _ = delete.insert(sentried_key.clone());
+                    }
+                };
+                (insert, update, delete)
+            },
+        );
+
+        self.hard_apply(operations)
+    }
+
+    fn hard_apply(&mut self, operations: SentriedOperations) -> Result<()> {
+        let (insert, update, delete) = operations;
```

Comment:
this usage is bit risky, especially `insert` and `update` has the same type.
can we make `SentriedOperations` as a proper struct and having a function returns insert/update/delete ?

Reply:
Yeah, that would seem better, I'll try something out for it.